File: src/j05-btr-distrib/btr.portal.web/src/navigation/portalMenuRegistry.ts
import { PortalMenuCodes } from './portalMenuCodes'
import type { PortalMenuGroup, PortalMenuGroupId, PortalMenuItem } from './portalMenu.types'

type PortalMenuItemInput = Omit<PortalMenuItem, 'isDashboard'> & { isDashboard?: boolean }

function item(entry: PortalMenuItemInput): PortalMenuItem {
  return {
    ...entry,
    isDashboard: entry.isDashboard ?? true,
  }
}

const portalMenuItems: PortalMenuItem[] = [
  item({
    code: PortalMenuCodes.EX01,
    label: 'Executive',
    icon: 'pi pi-th-large',
    routeName: 'dashboard',
    route: '/dashboard',
    order: 1,
    groupId: 'executive',
  }),
  item({
    code: PortalMenuCodes.EX02,
    label: 'Alert Center',
    icon: 'pi pi-bell',
    routeName: 'alert-center',
    route: '/alerts',
    order: 2,
    groupId: 'executive',
  }),
  item({
    code: PortalMenuCodes.EX03,
    label: 'Entity Analytics',
    icon: 'pi pi-id-card',
    routeName: 'entity-analytics-home',
    route: '/analytics',
    order: 3,
    groupId: 'executive',
    isDashboard: false,
  }),
  item({
    code: PortalMenuCodes.SA01,
    label: 'Sales',
    icon: 'pi pi-chart-line',
    routeName: 'sales-dashboard',
    route: '/dashboard/sales',
    order: 1,
    groupId: 'sales',
  }),
  item({
    code: PortalMenuCodes.SA02,
    label: 'Sales Forecast',
    icon: 'pi pi-chart-bar',
    routeName: 'sales-forecast-dashboard',
    route: '/dashboard/sales-forecast',
    order: 2,
    groupId: 'sales',
  }),
  item({
    code: PortalMenuCodes.SA03,
    label: 'Sales Report',
    icon: 'pi pi-list',
    routeName: 'sales-report',
    route: '/reports/sales',
    order: 3,
    groupId: 'sales',
    isDashboard: false,
  }),
  item({
    code: PortalMenuCodes.CU01,
    label: 'Customers',
    icon: 'pi pi-users',
    routeName: 'customers-dashboard',
    route: '/dashboard/customers',
    order: 1,
    groupId: 'customers',
  }),
  item({
    code: PortalMenuCodes.CU02,
    label: 'Customer Risk Forecast',
    icon: 'pi pi-shield',
    routeName: 'customer-risk-forecast-dashboard',
    route: '/dashboard/customer-risk-forecast',
    order: 2,
    groupId: 'customers',
  }),
  item({
    code: PortalMenuCodes.CU03,
    label: 'Collection Optimization',
    icon: 'pi pi-sliders-h',
    routeName: 'collection-optimization-dashboard',
    route: '/dashboard/collection-optimization',
    order: 3,
    groupId: 'customers',
  }),
  item({
    code: PortalMenuCodes.CU04,
    label: 'Customer Portfolio',
    icon: 'pi pi-briefcase',
    routeName: 'customer-portfolio-dashboard',
    route: '/dashboard/customer-portfolio',
    order: 4,
    groupId: 'customers',
  }),
  item({
    code: PortalMenuCodes.CU05,
    label: 'Customer Report',
    icon: 'pi pi-users',
    routeName: 'customer-report',
    route: '/reports/customers',
    order: 5,
    groupId: 'customers',
    isDashboard: false,
  }),
  item({
    code: PortalMenuCodes.FI01,
    label: 'Piutang',
    icon: 'pi pi-wallet',
    routeName: 'piutang-dashboard',
    route: '/dashboard/piutang',
    order: 1,
    groupId: 'finance',
  }),
  item({
    code: PortalMenuCodes.FI02,
    label: 'Collection',
    icon: 'pi pi-money-bill',
    routeName: 'collection-dashboard',
    route: '/dashboard/collection',
    order: 2,
    groupId: 'finance',
  }),
  item({
    code: PortalMenuCodes.FI03,
    label: 'Cash Flow Forecast',
    icon: 'pi pi-chart-line',
    routeName: 'cash-flow-forecast-dashboard',
    route: '/dashboard/cash-flow-forecast',
    order: 3,
    groupId: 'finance',
  }),
  item({
    code: PortalMenuCodes.FI04,
    label: 'Piutang Report',
    icon: 'pi pi-wallet',
    routeName: 'piutang-report',
    route: '/reports/piutang',
    order: 4,
    groupId: 'finance',
    isDashboard: false,
  }),
  item({
    code: PortalMenuCodes.SF01,
    label: 'Salesmen',
    icon: 'pi pi-id-card',
    routeName: 'salesmen-dashboard',
    route: '/dashboard/salesmen',
    order: 1,
    groupId: 'sales-force',
  }),
  item({
    code: PortalMenuCodes.SF02,
    label: 'Field Activity',
    icon: 'pi pi-map',
    routeName: 'field-activity-dashboard',
    route: '/dashboard/field-activity',
    order: 2,
    groupId: 'sales-force',
  }),
  item({
    code: PortalMenuCodes.IN01,
    label: 'Inventory',
    icon: 'pi pi-box',
    routeName: 'inventory-dashboard',
    route: '/dashboard/inventory',
    order: 1,
    groupId: 'inventory',
  }),
  item({
    code: PortalMenuCodes.IN02,
    label: 'Inventory Risk',
    icon: 'pi pi-exclamation-triangle',
    routeName: 'inventory-risk-dashboard',
    route: '/dashboard/inventory-risk',
    order: 2,
    groupId: 'inventory',
  }),
  item({
    code: PortalMenuCodes.IN03,
    label: 'Inventory Forecast',
    icon: 'pi pi-chart-line',
    routeName: 'inventory-forecast-dashboard',
    route: '/dashboard/inventory-forecast',
    order: 3,
    groupId: 'inventory',
  }),
  item({
    code: PortalMenuCodes.IN04,
    label: 'Inventory Optimization',
    icon: 'pi pi-sliders-h',
    routeName: 'inventory-optimization-dashboard',
    route: '/dashboard/inventory-optimization',
    order: 4,
    groupId: 'inventory',
  }),
  item({
    code: PortalMenuCodes.IN05,
    label: 'Inventory Report',
    icon: 'pi pi-box',
    routeName: 'inventory-report',
    route: '/reports/inventory',
    order: 5,
    groupId: 'inventory',
    isDashboard: false,
  }),
  item({
    code: PortalMenuCodes.PU01,
    label: 'Purchasing',
    icon: 'pi pi-shopping-cart',
    routeName: 'purchasing-dashboard',
    route: '/dashboard/purchasing',
    order: 1,
    groupId: 'purchasing',
  }),
  item({
    code: PortalMenuCodes.PU02,
    label: 'Purchasing Report',
    icon: 'pi pi-shopping-cart',
    routeName: 'purchasing-report',
    route: '/reports/purchasing',
    order: 2,
    groupId: 'purchasing',
    isDashboard: false,
  }),
  item({
    code: PortalMenuCodes.OP01,
    label: 'Locations',
    icon: 'pi pi-map-marker',
    routeName: 'locations-dashboard',
    route: '/dashboard/locations',
    order: 1,
    groupId: 'operations',
  }),
]

const groupDefinitions: Array<{ id: PortalMenuGroupId; label: string; order: number }> = [
  { id: 'executive', label: 'Executive', order: 1 },
  { id: 'sales', label: 'Sales', order: 2 },
  { id: 'customers', label: 'Customers', order: 3 },
  { id: 'finance', label: 'Finance', order: 4 },
  { id: 'sales-force', label: 'Sales Force', order: 5 },
  { id: 'inventory', label: 'Inventory', order: 6 },
  { id: 'purchasing', label: 'Purchasing', order: 7 },
  { id: 'operations', label: 'Operations', order: 8 },
]

export const portalMenuGroups: PortalMenuGroup[] = groupDefinitions.map((group) => ({
  ...group,
  items: portalMenuItems
    .filter((menuItem) => menuItem.groupId === group.id)
    .sort((a, b) => a.order - b.order),
}))

export const allPortalMenuItems: PortalMenuItem[] = [...portalMenuItems].sort((a, b) => {
  const groupA = groupDefinitions.find((g) => g.id === a.groupId)?.order ?? 99
  const groupB = groupDefinitions.find((g) => g.id === b.groupId)?.order ?? 99
  if (groupA !== groupB) return groupA - groupB
  return a.order - b.order
})

export function getDomainDashboardLinks(): PortalMenuItem[] {
  return allPortalMenuItems.filter(
    (menuItem) => menuItem.isDashboard && menuItem.code !== PortalMenuCodes.EX02,
  )
}

File: src/j05-btr-distrib/btr.application/ReportingContext/Shared/PortalMenuRegistry.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace btr.application.ReportingContext.Shared
{
    public sealed class PortalMenuLinkModel
    {
        public PortalMenuLinkModel(string code, string label, string route)
        {
            Code = code;
            Label = label;
            Route = route;
        }

        public string Code { get; }

        public string Label { get; }

        public string Route { get; }
    }

    internal sealed class PortalMenuEntry
    {
        public PortalMenuEntry(
            string code,
            string label,
            string route,
            int groupOrder,
            int itemOrder,
            bool isDashboard = true)
        {
            Code = code;
            Label = label;
            Route = route;
            GroupOrder = groupOrder;
            ItemOrder = itemOrder;
            IsDashboard = isDashboard;
        }

        public string Code { get; }

        public string Label { get; }

        public string Route { get; }

        public int GroupOrder { get; }

        public int ItemOrder { get; }

        public bool IsDashboard { get; }
    }

    public static class PortalMenuRegistry
    {
        public const string AlertCenterCode = "EX02";

        private static readonly IReadOnlyList<PortalMenuEntry> Entries = BuildEntries();

        public static IReadOnlyList<PortalMenuLinkModel> GetAllLinks() =>
            Entries
                .OrderBy(e => e.GroupOrder)
                .ThenBy(e => e.ItemOrder)
                .Select(ToLink)
                .ToList();

        public static IReadOnlyList<PortalMenuLinkModel> GetDomainDashboardLinks() =>
            Entries
                .Where(e => e.IsDashboard && !string.Equals(e.Code, AlertCenterCode, StringComparison.Ordinal))
                .OrderBy(e => e.GroupOrder)
                .ThenBy(e => e.ItemOrder)
                .Select(ToLink)
                .ToList();

        public static string FormatMenuLabel(string route, string fallbackLabel = null)
        {
            if (TryGetByRoute(route, out var entry))
                return $"{entry.Code} · {entry.Label}";

            return fallbackLabel ?? route;
        }

        public static bool TryGetByRoute(string route, out PortalMenuLinkModel link)
        {
            link = null;
            if (!TryGetEntryByRoute(route, out var entry))
                return false;

            link = ToLink(entry);
            return true;
        }

        private static bool TryGetEntryByRoute(string route, out PortalMenuEntry entry)
        {
            entry = Entries.FirstOrDefault(e =>
                string.Equals(NormalizeRoute(e.Route), NormalizeRoute(route), StringComparison.OrdinalIgnoreCase));
            return entry != null;
        }

        private static PortalMenuLinkModel ToLink(PortalMenuEntry entry) =>
            new PortalMenuLinkModel(entry.Code, entry.Label, entry.Route);

        private static string NormalizeRoute(string route)
        {
            if (string.IsNullOrWhiteSpace(route))
                return route;

            return route.EndsWith("/", StringComparison.Ordinal) && route.Length > 1
                ? route.TrimEnd('/')
                : route;
        }

        private static IReadOnlyList<PortalMenuEntry> BuildEntries()
        {
            return new List<PortalMenuEntry>
            {
                Entry("EX01", "Executive", "/dashboard", 1, 1),
                Entry("EX02", "Alert Center", "/alerts", 1, 2),
                Entry("SA01", "Sales", "/dashboard/sales", 2, 1),
                Entry("SA02", "Sales Forecast", "/dashboard/sales-forecast", 2, 2),
                Entry("SA03", "Sales Report", "/reports/sales", 2, 3, isDashboard: false),
                Entry("CU01", "Customers", "/dashboard/customers", 3, 1),
                Entry("CU02", "Customer Risk Forecast", "/dashboard/customer-risk-forecast", 3, 2),
                Entry("CU03", "Collection Optimization", "/dashboard/collection-optimization", 3, 3),
                Entry("CU04", "Customer Portfolio", "/dashboard/customer-portfolio", 3, 4),
                Entry("CU05", "Customer Report", "/reports/customers", 3, 5, isDashboard: false),
                Entry("FI01", "Piutang", "/dashboard/piutang", 4, 1),
                Entry("FI02", "Collection", "/dashboard/collection", 4, 2),
                Entry("FI03", "Cash Flow Forecast", "/dashboard/cash-flow-forecast", 4, 3),
                Entry("FI04", "Piutang Report", "/reports/piutang", 4, 4, isDashboard: false),
                Entry("SF01", "Salesmen", "/dashboard/salesmen", 5, 1),
                Entry("SF02", "Sales Force Overview", "/dashboard/field-activity", 5, 2),
                Entry("SF03", "Salesman Field Activity", "/dashboard/field-activity/detail", 5, 3),
                Entry("IN01", "Inventory", "/dashboard/inventory", 6, 1),
                Entry("IN02", "Inventory Risk", "/dashboard/inventory-risk", 6, 2),
                Entry("IN03", "Inventory Forecast", "/dashboard/inventory-forecast", 6, 3),
                Entry("IN04", "Inventory Optimization", "/dashboard/inventory-optimization", 6, 4),
                Entry("IN05", "Inventory Report", "/reports/inventory", 6, 5, isDashboard: false),
                Entry("PU01", "Purchasing", "/dashboard/purchasing", 7, 1),
                Entry("PU02", "Purchasing Report", "/reports/purchasing", 7, 2, isDashboard: false),
                Entry("OP01", "Locations", "/dashboard/locations", 8, 1),
            };
        }

        private static PortalMenuEntry Entry(
            string code,
            string label,
            string route,
            int groupOrder,
            int itemOrder,
            bool isDashboard = true) =>
            new PortalMenuEntry(code, label, route, groupOrder, itemOrder, isDashboard);
    }
}
